.github/workflows/create-packages.yml

```diff
@@ -0,0 +1,81 @@
+name: Create package from source tarball
+
+
+on:
+  workflow_call:
+    inputs:
+      source-tarball-artifact-name:
+        required: true
+        type: string
+      dbld-image-mode:
+        required: true
+        type: string  # cache / build
+      distros:
+        required: false
+        type: string
+        default: '[
+          "debian-stretch",
+          "debian-buster",
+          "debian-testing",
+          "debian-sid",
+          "ubuntu-xenial",
+          "ubuntu-bionic",
+          "debian-bullseye",
+          "ubuntu-focal",
+          "ubuntu-impish",
+          "centos-7",
+          "fedora-33"
+        ]'
+
+
+jobs:
+  create-packages:
+    name: ${{ matrix.distro }}
+
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        distro: ${{ fromJson(inputs.distros) }}
+      fail-fast: false
+
+    steps:
+      - name: Download source tarball artifact
+        uses: actions/download-artifact@v2
+        with:
+          name: ${{ inputs.source-tarball-artifact-name }}
+
+      - name: Extract source tarball
+        run: |
+          mkdir syslog-ng
+          tar --strip-components=1 -xvf syslog-ng*.tar.gz -C syslog-ng
+
+      - name: Prepare docker image
+        working-directory: syslog-ng
+        run: |
+          if [[ "${{ inputs.dbld-image-mode }}" = "build" ]]
+          then
+            ./dbld/rules image-${{ matrix.distro }}
+          elif [[ "${{ inputs.dbld-image-mode }}" = "cache" ]]
+          then
+            ./dbld/rules cache-image-${{ matrix.distro }}
+          else
+            echo Unexpected input: dbld-image-mode=${{ inputs.dbld-image-mode }}
+            false
+          fi
+
+      - name: Create package
+        working-directory: syslog-ng
+        run: |
+          ./dbld/rules package-${{ matrix.distro }}
+
+      - name: Prepare package for artifact
+        run: |
+          mkdir package
+          cp -r syslog-ng/dbld/build/${{ matrix.distro }} package/
```

Comment:
I think it is necessary, but maybe I am wrong.

`upload-artifact` "flattens" the path, so `path: /home/alltilla/whatever` would upload the contents of `whatever`, but not the directory itself.

I wanted to keep the `whatever` directory in the artifact, but `/home/alltilla` has other directories as well, so I could not write `/home/alltilla/*`, so I created a new, empty directory (e.g. `/home/alltilla/foo`), and copied `whatever` there, so `path: /home/alltilla/foo/*` will upload the `whatever` directory with its contents.

I would prefer if it was not needed, maybe others can find a simpler solution. I will check it again, though!

Reply:
Thanks for the explanation!
I've checked our current package uploads (e.g. [here](https://github.com/syslog-ng/syslog-ng/actions/runs/2000844525) is a run), and this one, and yes, the directory is not kept.

I don't think it's needed to create unique directories for artifacts, that's solved by Github Actions if you give a **unique name** for artifact upload.
Maybe I don't understand the motivation behind keeping the directory. You can avoid overwriting previously downloaded artifacts with a unique `path` when downloading.

Reply:
Okay, I found where is the directory name of the packages (e.g. "centos-7) needed (at leas now):
In `upload-packages.yml`, the job first downloads the given package by the name `package-centos-7` and then when it's about to upload the packages to Azure, it **knows** that there is a `centos-7` folder (by using `${{ matrix.distro }}` context variable.
It's confusing that the distro variable is used for the source parameter: it is not related to the list in the `distros` input variable, it is a knowledge from the `create-packages` job that a certain directory will be created.
I wouldn't know it internal knowledge though, as you can expect the content of an artifact, but if we would change something in `create-packages` you could break `upload-packages` if it's not tested.
In my opinion it would be easier if we add a path to the `download-artifact` step and then use that directory for uploading to azure.

Reply:
Good idea, thanks. I am trying to make the paths more simpler with your idea.

Reply:
Wow it got a lot simpler with your idea, thanks! :)

Reply:
Thanks for the change!

I would accept this version, but I would like to hear your opinion first. I've unresolved this thread for visibility reasons only.

I like the current version however, I think the additional `Prepare package for artifact` step could be skipped. 
I think it's unnecessary and the extra directory serves no meaning.

IMHO, We could simplify it:
```
      - name: Create package
        working-directory: syslog-ng
        run: |
          ./dbld/rules package-${{ matrix.distro }}

      - name: Store package as artifact
        uses: actions/upload-artifact@v3
        with:
          name: package-${{ matrix.distro }}
          path: syslog-ng/dbld/build/${{ matrix.distro }}
```

`upload-packages` workflow would work without changes, as it only cares about the name of the artifact name, and it doesn't care much about the internal structure of the artifact: it just downloads it to a given path, then uploads it to Azure.

Reply:
I think the problem with your solution is that when you explicitly give a directory path for `upload-artifact`, it will not upload the directory and its contents, but only its contents (or at least this is how I remember). I will check it again though, it would be nice to remove the preparation step.

Reply:
Is it necessary to have a directory, though? The `upload-packages` step will download the artifact into a separate directory, why need another directory inside that?

Reply:
When we upload to Azure, we need to put the deb files for example to `incoming/nightly/debian-buster/...deb`. So we need to have the `debian-buster` directory path somewhere.

The `azure-cli` step in `upload-packages` needs a `--source` (local) and a `--destination` (remote) parameter, and it will match the contents of the two. The way it works now is that the source is `package` (where the artifact is downloaded, currently has the `debian-buster` dir inside) and the destination is `/incoming/nightly`. I think it is clean this way, `upload-artifact` does not make tricks with the paths, it simply uploads the content of the artifact to `incoming/nightly/` for example. It would be possible to add `debian-buster` to the end of `--destination`, but then it would introduce some unnecesary logic to `upload-packages`, which might be better to have in `create-packages` instead.

I think having the necessary directory structure in the artifact itself, and not delegating this logic to `upload-packages` is cleaner in a high level perspective. Low level it introduces a "hacky" step, which is needed because of the way `upload-artifact` works. I would prefer this way.

Reply:
Updated.

Thanks for the discussion :)